### PR TITLE
experimental: output clean OSS-Fuzz & CFLite folders

### PR DESCRIPTION
**Pull request by DavidKorczynski**
This adds output from the build/evaluator to create clean OSS-Fuzz/CFLite projects for each successful harness generated. These can be copied over to OSS-Fuzz and run without issues.

### CODE REVIEW

File: experimental/c-cpp/manager.py [hunk 1]
```diff
@@ -772,6 +852,96 @@ def evaluate_heuristic(test_dir, result_to_validate, fuzzer_intrinsics,
...
+  with open(os.path.join(cflite_folder, 'project.yaml'), 'w') as project_out:
+    yaml.dump(project_yaml, project_out)
+
+  # Copy fuzzer
+  shutil.copy(os.path.join(success_dir, 'empty-fuzzer.cpp'),
+              os.path.join(cflite_folder, 'fuzzer.cpp'))
+
+  # Create Dockerfile
+  project_repo_dir = github_repo.split('/')[-1]
+  dockerfile = CLEAN_DOCKER_CFLITE % (project_repo_dir, project_repo_dir)
```
Comment (oliverchang) [inline experimental/c-cpp/manager.py:922]: nit: can we use str.format here with named arguments here instead?

File: experimental/c-cpp/manager.py [hunk 2]
```diff
@@ -772,6 +852,96 @@ def evaluate_heuristic(test_dir, result_to_validate, fuzzer_intrinsics,
...
+  with open(os.path.join(success_dir, 'build.sh'), 'r') as f:
+    build_content = f.read()
+  split_build_content = build_content.split("\n")
+
+  original_build_folder = split_build_content[1].split('/')[-1]
+
+  clean_build_content_lines = split_build_content[:1] + split_build_content[4:]
+  clean_build_content = '\n'.join(clean_build_content_lines).replace(
+      '/src/generated-fuzzer',
+      '$OUT/fuzzer').replace('empty-fuzzer', 'fuzzer').replace(
```
Comment (oliverchang) [inline experimental/c-cpp/manager.py:936]: Can you add some comments explaining these replacements a bit more?

**Review by oliverchang — APPROVED** [review]

File: experimental/c-cpp/manager.py [hunk 3]
```diff
@@ -772,6 +852,96 @@ def evaluate_heuristic(test_dir, result_to_validate, fuzzer_intrinsics,
...
+          '/src/', '$SRC/').replace(original_build_folder, project_repo_dir)
```
Comment (DonggeLiu) [inline experimental/c-cpp/manager.py:898]: A silly question:
What is the goal of modifying the `build.sh` here?
In particular, why can we assume the line numbers (`1`, `4`) in this line?
```python
clean_build_content_lines = split_build_content[:1] + split_build_content[4:]
```

File: experimental/c-cpp/manager.py [hunk 4]
```diff
@@ -772,6 +852,96 @@ def evaluate_heuristic(test_dir, result_to_validate, fuzzer_intrinsics,
...
+  # Create Dockerfile
+  project_repo_dir = github_repo.split('/')[-1]
+  dockerfile = CLEAN_DOCKER_CFLITE % (project_repo_dir, project_repo_dir)
+  with open(os.path.join(cflite_folder, 'Dockerfile'), 'w') as docker_out:
+    docker_out.write(dockerfile)
+
+  # Build file
+  with open(os.path.join(success_dir, 'build.sh'), 'r') as f:
+    build_content = f.read()
+  split_build_content = build_content.split("\n")
```
Comment (DonggeLiu) [inline experimental/c-cpp/manager.py:929]: nit: `'\n'` as above.

File: experimental/c-cpp/manager.py (hunk 3 above)
Comment (DavidKorczynski) [reply experimental/c-cpp/manager.py]: The goal is to convert it from how it looks during auto-gen testing/evaluation stages to be something that can work proper with a clean OSS-Fuzz project.

Build used during testing inside of base-builder container:

```sh
#!/bin/bash
rm -rf //src/test-fuzz-build-7
cp -rf /src/project_source /src/test-fuzz-build-7
cd /src/test-fuzz-build-7
mkdir fuzz-build
cd fuzz-build
cmake -DCMAKE_VERBOSE_MAKEFILE=ON ../
make V=1 || true
```

Build for clean OSS-Fuzz project

```sh
#!/bin/bash
mkdir fuzz-build
cd fuzz-build
cmake -DCMAKE_VERBOSE_MAKEFILE=ON ../
make V=1 || true
```

The lines you highlight cut out the following from the build script:

```sh
rm -rf //src/test-fuzz-build-7
cp -rf /src/project_source /src/test-fuzz-build-7
cd /src/test-fuzz-build-7
```

File: experimental/c-cpp/manager.py (hunk 1 above)
Comment (DavidKorczynski) [reply experimental/c-cpp/manager.py]: Yes, done

File: experimental/c-cpp/manager.py (hunk 2 above)
Comment (DavidKorczynski) [reply experimental/c-cpp/manager.py:954]: Yes, done.

File: experimental/c-cpp/manager.py [hunk 5]
```diff
@@ -54,6 +54,86 @@
     return 0;
 }'''
 
+CLEAN_DOCKER = '''# Copyright 2024 Google LLC
```
[the remark this answers is not in the document]
Comment (DavidKorczynski) [reply experimental/c-cpp/manager.py:57]: I would prefer to keep it as is for now -- maybe change this in the longer run once things have settled a bit with this experimental logic. At the moment it's quite convenient to work with a single file, e.g. for moving it in and out of containers easily.

File: experimental/c-cpp/manager.py [hunk 6]
```diff
@@ -772,6 +852,96 @@ def evaluate_heuristic(test_dir, result_to_validate, fuzzer_intrinsics,
...
+  # Create Dockerfile
+  project_repo_dir = github_repo.split('/')[-1]
+  dockerfile = CLEAN_DOCKER % (github_repo, project_repo_dir, project_repo_dir)
+  with open(os.path.join(oss_fuzz_folder, 'Dockerfile'), 'w') as docker_out:
+    docker_out.write(dockerfile)
+
+  # Build file
+  with open(os.path.join(success_dir, 'build.sh'), 'r') as f:
+    build_content = f.read()
+  split_build_content = build_content.split("\n")
```
[the remark this answers is not in the document]
Comment (DavidKorczynski) [reply experimental/c-cpp/manager.py]: Thanks, done

File: experimental/c-cpp/manager.py (hunk 4 above)
Comment (DavidKorczynski) [reply experimental/c-cpp/manager.py]: Thanks, done

File: experimental/c-cpp/manager.py [hunk 7]
```diff
@@ -54,6 +54,86 @@
...
+# limitations under the License.
+#
+################################################################################
+
+FROM gcr.io/oss-fuzz-base/base-builder
+RUN apt-get update && apt-get install -y make autoconf automake libtool cmake \\
+                      pkg-config curl check
+COPY *.sh *.cpp $SRC/
+RUN git clone %s %s
+WORKDIR %s'''
```
[the remark this answers is not in the document]
Comment (DavidKorczynski) [reply experimental/c-cpp/manager.py]: Yes, done

**Review by DonggeLiu — APPROVED** [review]
Thanks for the explanations and addressing the comments, @DavidKorczynski .
They LGTM except one tiny thing I just noticed:
Some functions in this file have type-hinting (which is great and consistent with the rest files in this repo) but others do not.

Could you add type-hinting to them?
If you have other tasks in hand, this is not urgent because they are in `experimental/` but I guess it is easier to add them before memory decays?
Thanks

**Comment by DavidKorczynski** [comment]
> Could you add type-hinting to them?

Yep, done!

**Pull request merged**